expression/builtin_convert_charset.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// InternalFuncToBinary accepts a string and returns another string encoded in a given charset.
+const InternalFuncToBinary = "to_binary"
+
+type tidbConvertCharsetFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *tidbConvertCharsetFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, c.verifyArgs(args)
+	}
+	argFieldTp := args[0].GetType()
+	argTp := argFieldTp.EvalType()
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, argTp, argTp)
+	if err != nil {
+		return nil, err
+	}
+	var sig builtinFunc
+	switch argTp {
+	case types.ETString:
+		sig = &builtinInternalToBinarySig{bf}
+		// TODO(tangenta): set pbcode for to_binary() sig.
+		sig.setPbCode(tipb.ScalarFuncSig_Unspecified)
+	default:
+		msg := fmt.Sprintf("unexpected argTp: %d", argTp)
+		panic(msg)
+	}
+	return sig, nil
+}
+
+var _ builtinFunc = &builtinInternalToBinarySig{}
+
+type builtinInternalToBinarySig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinInternalToBinarySig) Clone() builtinFunc {
+	newSig := &builtinInternalToBinarySig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinInternalToBinarySig) evalString(row chunk.Row) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return res, isNull, err
+	}
+	tp := b.args[0].GetType()
+	enc := charset.NewEncoding(tp.Charset)
+	res, err = enc.EncodeString(val)
+	return res, false, err
+}
+
+func (b *builtinInternalToBinarySig) vectorized() bool {
+	return true
+}
+
+func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get()
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	enc := charset.NewEncoding(b.args[0].GetType().Charset)
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		var str string
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		str = buf.GetString(i)
+		str, err = enc.EncodeString(str)
+		if err != nil {
+			return err
+		}
+		result.AppendString(str)
+	}
+	return nil
+}
+
+// toBinaryMap contains the builtin functions which arguments need to be converted to the correct charset.
+var toBinaryMap = map[string]struct{}{
+	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
+	ast.ToBase64: {},
+}
+
+// WrapWithToBinary wraps `expr` with to_binary sig.
+func WrapWithToBinary(ctx sessionctx.Context, expr Expression, funcName string) Expression {
+	retTp := expr.GetType()
+	if _, err := charset.GetDefaultCollationLegacy(retTp.Charset); err != nil {
+		if _, ok := toBinaryMap[funcName]; ok {
+			bf, err := newBaseBuiltinFunc(ctx, InternalFuncToBinary, []Expression{expr}, retTp.EvalType())
```

Comment:
maybe we can use `getFunction`, it will also set pbcode

Reply:
Done.